## 实习四：基于SQL实现机器学习的基本概念
> 成员：罗逸龙2000017781  占可盈2000013136  范居令2000012942 


In [65]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [66]:
import pymysql 
pymysql.install_as_MySQLdb()
%sql mysql://stu2000017781:stu2000017781@162.105.146.37:43306

In [67]:
%sql use stu2000017781;

 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.


[]

### 练习一：发现数据中隐含的辛普森悖论
#### 任务一：使用纯SQL生成报表，并判断是否存在辛普森悖论

首先需要根据要求生成数据集

In [68]:
%%sql

DROP TABLE IF EXISTS application;

CREATE TABLE application (
  id INT,
  sex VARCHAR(10),
  department VARCHAR(20),
  acception VARCHAR(10)
);

INSERT INTO application (id, sex, department, acception)
VALUES
    (1, 'Female', 'Business School', 'Accepted'),
    (2, 'Female', 'Business School', 'Accepted'),
    (3, 'Female', 'Business School', 'Accepted'),
    (4, 'Female', 'Business School', 'Accepted'),
    (5, 'Female', 'Business School', 'Accepted'),
    (6, 'Female', 'Business School', 'Accepted'),
    (7, 'Female', 'Business School', 'Accepted'),
    (8, 'Female', 'Business School', 'Accepted'),
    (9, 'Female', 'Business School', 'Accepted'),
    (10, 'Female', 'Business School', 'Accepted'),
    (11, 'Female', 'Business School', 'Accepted'),
    (12, 'Female', 'Business School', 'Accepted'),
    (13, 'Female', 'Business School', 'Accepted'),
    (14, 'Female', 'Business School', 'Accepted'),
    (15, 'Female', 'Business School', 'Accepted'),
    (16, 'Female', 'Business School', 'Accepted'),
    (17, 'Female', 'Business School', 'Accepted'),
    (18, 'Female', 'Business School', 'Accepted'),
    (19, 'Female', 'Business School', 'Accepted'),
    (20, 'Female', 'Business School', 'Accepted'),
    (21, 'Female', 'Business School', 'Accepted'),
    (22, 'Female', 'Business School', 'Accepted'),
    (23, 'Female', 'Business School', 'Accepted'),
    (24, 'Female', 'Business School', 'Accepted'),
    (25, 'Female', 'Business School', 'Accepted'),
    (26, 'Female', 'Business School', 'Accepted'),
    (27, 'Female', 'Business School', 'Accepted'),
    (28, 'Female', 'Business School', 'Accepted'),
    (29, 'Female', 'Business School', 'Accepted'),
    (30, 'Female', 'Business School', 'Accepted'),
    (31, 'Female', 'Business School', 'Accepted'),
    (32, 'Female', 'Business School', 'Accepted'),
    (33, 'Female', 'Business School', 'Accepted'),
    (34, 'Female', 'Business School', 'Accepted'),
    (35, 'Female', 'Business School', 'Accepted'),
    (36, 'Female', 'Business School', 'Accepted'),
    (37, 'Female', 'Business School', 'Accepted'),
    (38, 'Female', 'Business School', 'Accepted'),
    (39, 'Female', 'Business School', 'Accepted'),
    (40, 'Female', 'Business School', 'Accepted'),
    (41, 'Female', 'Business School', 'Accepted'),
    (42, 'Female', 'Business School', 'Accepted'),
    (43, 'Female', 'Business School', 'Accepted'),
    (44, 'Female', 'Business School', 'Accepted'),
    (45, 'Female', 'Business School', 'Accepted'),
    (46, 'Female', 'Business School', 'Accepted'),
    (47, 'Female', 'Business School', 'Accepted'),
    (48, 'Female', 'Business School', 'Accepted'),
    (49, 'Female', 'Business School', 'Accepted'),
    (50, 'Female', 'Business School', 'Rejected'),
    (51, 'Female', 'Business School', 'Rejected'),
    (52, 'Female', 'Business School', 'Rejected'),
    (53, 'Female', 'Business School', 'Rejected'),
    (54, 'Female', 'Business School', 'Rejected'),
    (55, 'Female', 'Business School', 'Rejected'),
    (56, 'Female', 'Business School', 'Rejected'),
    (57, 'Female', 'Business School', 'Rejected'),
    (58, 'Female', 'Business School', 'Rejected'),
    (59, 'Female', 'Business School', 'Rejected'),
    (60, 'Female', 'Business School', 'Rejected'),
    (61, 'Female', 'Business School', 'Rejected'),
    (62, 'Female', 'Business School', 'Rejected'),
    (63, 'Female', 'Business School', 'Rejected'),
    (64, 'Female', 'Business School', 'Rejected'),
    (65, 'Female', 'Business School', 'Rejected'),
    (66, 'Female', 'Business School', 'Rejected'),
    (67, 'Female', 'Business School', 'Rejected'),
    (68, 'Female', 'Business School', 'Rejected'),
    (69, 'Female', 'Business School', 'Rejected'),
    (70, 'Female', 'Business School', 'Rejected'),
    (71, 'Female', 'Business School', 'Rejected'),
    (72, 'Female', 'Business School', 'Rejected'),
    (73, 'Female', 'Business School', 'Rejected'),
    (74, 'Female', 'Business School', 'Rejected'),
    (75, 'Female', 'Business School', 'Rejected'),
    (76, 'Female', 'Business School', 'Rejected'),
    (77, 'Female', 'Business School', 'Rejected'),
    (78, 'Female', 'Business School', 'Rejected'),
    (79, 'Female', 'Business School', 'Rejected'),
    (80, 'Female', 'Business School', 'Rejected'),
    (81, 'Female', 'Business School', 'Rejected'),
    (82, 'Female', 'Business School', 'Rejected'),
    (83, 'Female', 'Business School', 'Rejected'),
    (84, 'Female', 'Business School', 'Rejected'),
    (85, 'Female', 'Business School', 'Rejected'),
    (86, 'Female', 'Business School', 'Rejected'),
    (87, 'Female', 'Business School', 'Rejected'),
    (88, 'Female', 'Business School', 'Rejected'),
    (89, 'Female', 'Business School', 'Rejected'),
    (90, 'Female', 'Business School', 'Rejected'),
    (91, 'Female', 'Business School', 'Rejected'),
    (92, 'Female', 'Business School', 'Rejected'),
    (93, 'Female', 'Business School', 'Rejected'),
    (94, 'Female', 'Business School', 'Rejected'),
    (95, 'Female', 'Business School', 'Rejected'),
    (96, 'Female', 'Business School', 'Rejected'),
    (97, 'Female', 'Business School', 'Rejected'),
    (98, 'Female', 'Business School', 'Rejected'),
    (99, 'Female', 'Business School', 'Rejected'),
    (100, 'Female', 'Business School', 'Rejected');

INSERT INTO application (id, sex, department, acception)
VALUES
    (101, 'Male', 'Business School', 'Accepted'),
    (102, 'Male', 'Business School', 'Accepted'),
    (103, 'Male', 'Business School', 'Accepted'),
    (104, 'Male', 'Business School', 'Accepted'),
    (105, 'Male', 'Business School', 'Accepted'),
    (106, 'Male', 'Business School', 'Accepted'),
    (107, 'Male', 'Business School', 'Accepted'),
    (108, 'Male', 'Business School', 'Accepted'),
    (109, 'Male', 'Business School', 'Accepted'),
    (110, 'Male', 'Business School', 'Accepted'),
    (111, 'Male', 'Business School', 'Accepted'),
    (112, 'Male', 'Business School', 'Accepted'),
    (113, 'Male', 'Business School', 'Accepted'),
    (114, 'Male', 'Business School', 'Accepted'),
    (115, 'Male', 'Business School', 'Accepted'),
    (116, 'Male', 'Business School', 'Rejected'),
    (117, 'Male', 'Business School', 'Rejected'),
    (118, 'Male', 'Business School', 'Rejected'),
    (119, 'Male', 'Business School', 'Rejected'),
    (120, 'Male', 'Business School', 'Rejected');
    
INSERT INTO application (id, sex, department, acception)
VALUES
    (121, 'Female', 'Law School', 'Accepted'),
    (122, 'Female', 'Law School', 'Rejected'),
    (123, 'Female', 'Law School', 'Rejected'),
    (124, 'Female', 'Law School', 'Rejected'),
    (125, 'Female', 'Law School', 'Rejected'),
    (126, 'Female', 'Law School', 'Rejected'),
    (127, 'Female', 'Law School', 'Rejected'),
    (128, 'Female', 'Law School', 'Rejected'),
    (129, 'Female', 'Law School', 'Rejected'),
    (130, 'Female', 'Law School', 'Rejected'),
    (131, 'Female', 'Law School', 'Rejected'),
    (132, 'Female', 'Law School', 'Rejected'),
    (133, 'Female', 'Law School', 'Rejected'),
    (134, 'Female', 'Law School', 'Rejected'),
    (135, 'Female', 'Law School', 'Rejected'),
    (136, 'Female', 'Law School', 'Rejected'),
    (137, 'Female', 'Law School', 'Rejected'),
    (138, 'Female', 'Law School', 'Rejected'),
    (139, 'Female', 'Law School', 'Rejected'),
    (140, 'Female', 'Law School', 'Rejected');

-- 插入法学院男生申请数据
INSERT INTO application (id, sex, department, acception)
VALUES
    (141, 'Male', 'Law School', 'Accepted'),
    (142, 'Male', 'Law School', 'Accepted'),
    (143, 'Male', 'Law School', 'Accepted'),
    (144, 'Male', 'Law School', 'Accepted'),
    (145, 'Male', 'Law School', 'Accepted'),
    (146, 'Male', 'Law School', 'Accepted'),
    (147, 'Male', 'Law School', 'Accepted'),
    (148, 'Male', 'Law School', 'Accepted'),
    (149, 'Male', 'Law School', 'Accepted'),
    (150, 'Male', 'Law School', 'Accepted'),
    (151, 'Male', 'Law School', 'Rejected'),
    (152, 'Male', 'Law School', 'Rejected'),
    (153, 'Male', 'Law School', 'Rejected'),
    (154, 'Male', 'Law School', 'Rejected'),
    (155, 'Male', 'Law School', 'Rejected'),
    (156, 'Male', 'Law School', 'Rejected'),
    (157, 'Male', 'Law School', 'Rejected'),
    (158, 'Male', 'Law School', 'Rejected'),
    (159, 'Male', 'Law School', 'Rejected'),
    (160, 'Male', 'Law School', 'Rejected'),
    (161, 'Male', 'Law School', 'Rejected'),
    (162, 'Male', 'Law School', 'Rejected'),
    (163, 'Male', 'Law School', 'Rejected'),
    (164, 'Male', 'Law School', 'Rejected'),
    (165, 'Male', 'Law School', 'Rejected'),
    (166, 'Male', 'Law School', 'Rejected'),
    (167, 'Male', 'Law School', 'Rejected'),
    (168, 'Male', 'Law School', 'Rejected'),
    (169, 'Male', 'Law School', 'Rejected'),
    (170, 'Male', 'Law School', 'Rejected'),
    (171, 'Male', 'Law School', 'Rejected'),
    (172, 'Male', 'Law School', 'Rejected'),
    (173, 'Male', 'Law School', 'Rejected'),
    (174, 'Male', 'Law School', 'Rejected'),
    (175, 'Male', 'Law School', 'Rejected'),
    (176, 'Male', 'Law School', 'Rejected'),
    (177, 'Male', 'Law School', 'Rejected'),
    (178, 'Male', 'Law School', 'Rejected'),
    (179, 'Male', 'Law School', 'Rejected'),
    (180, 'Male', 'Law School', 'Rejected'),
    (181, 'Male', 'Law School', 'Rejected'),
    (182, 'Male', 'Law School', 'Rejected'),
    (183, 'Male', 'Law School', 'Rejected'),
    (184, 'Male', 'Law School', 'Rejected'),
    (185, 'Male', 'Law School', 'Rejected'),
    (186, 'Male', 'Law School', 'Rejected'),
    (187, 'Male', 'Law School', 'Rejected'),
    (188, 'Male', 'Law School', 'Rejected'),
    (189, 'Male', 'Law School', 'Rejected'),
    (190, 'Male', 'Law School', 'Rejected'),
    (191, 'Male', 'Law School', 'Rejected'),
    (192, 'Male', 'Law School', 'Rejected'),
    (193, 'Male', 'Law School', 'Rejected'),
    (194, 'Male', 'Law School', 'Rejected'),
    (195, 'Male', 'Law School', 'Rejected'),
    (196, 'Male', 'Law School', 'Rejected'),
    (197, 'Male', 'Law School', 'Rejected'),
    (198, 'Male', 'Law School', 'Rejected'),
    (199, 'Male', 'Law School', 'Rejected'),
    (200, 'Male', 'Law School', 'Rejected'),
    (201, 'Male', 'Law School', 'Rejected'),
    (202, 'Male', 'Law School', 'Rejected'),
    (203, 'Male', 'Law School', 'Rejected'),
    (204, 'Male', 'Law School', 'Rejected'),
    (205, 'Male', 'Law School', 'Rejected'),
    (206, 'Male', 'Law School', 'Rejected'),
    (207, 'Male', 'Law School', 'Rejected'),
    (208, 'Male', 'Law School', 'Rejected'),
    (209, 'Male', 'Law School', 'Rejected'),
    (210, 'Male', 'Law School', 'Rejected'),
    (211, 'Male', 'Law School', 'Rejected'),
    (212, 'Male', 'Law School', 'Rejected'),
    (213, 'Male', 'Law School', 'Rejected'),
    (214, 'Male', 'Law School', 'Rejected'),
    (215, 'Male', 'Law School', 'Rejected'),
    (216, 'Male', 'Law School', 'Rejected'),
    (217, 'Male', 'Law School', 'Rejected'),
    (218, 'Male', 'Law School', 'Rejected'),
    (219, 'Male', 'Law School', 'Rejected'),
    (220, 'Male', 'Law School', 'Rejected'),
    (221, 'Male', 'Law School', 'Rejected'),
    (222, 'Male', 'Law School', 'Rejected'),
    (223, 'Male', 'Law School', 'Rejected'),
    (224, 'Male', 'Law School', 'Rejected'),
    (225, 'Male', 'Law School', 'Rejected'),
    (226, 'Male', 'Law School', 'Rejected'),
    (227, 'Male', 'Law School', 'Rejected'),
    (228, 'Male', 'Law School', 'Rejected'),
    (229, 'Male', 'Law School', 'Rejected'),
    (230, 'Male', 'Law School', 'Rejected'),
    (231, 'Male', 'Law School', 'Rejected'),
    (232, 'Male', 'Law School', 'Rejected'),
    (233, 'Male', 'Law School', 'Rejected'),
    (234, 'Male', 'Law School', 'Rejected'),
    (235, 'Male', 'Law School', 'Rejected'),
    (236, 'Male', 'Law School', 'Rejected'),
    (237, 'Male', 'Law School', 'Rejected'),
    (238, 'Male', 'Law School', 'Rejected'),
    (239, 'Male', 'Law School', 'Rejected'),
    (240, 'Male', 'Law School', 'Rejected');

 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.
0 rows affected.
100 rows affected.
20 rows affected.
20 rows affected.
100 rows affected.


[]

In [69]:
%%sql 

SELECT 
    '商学院' AS 学院,
    COUNT(CASE WHEN sex = 'Female' THEN 1 END) AS 女生申请人数,
    COUNT(CASE WHEN sex = 'Female' AND acception = 'Accepted' THEN 1 END) AS 女生录取人数,
    ROUND(COUNT(CASE WHEN sex = 'Female' AND acception = 'Accepted' THEN 1 END) / COUNT(CASE WHEN sex = 'Female' THEN 1 END) * 100, 2) AS 女生录取率,
    COUNT(CASE WHEN sex = 'Male' THEN 1 END) AS 男生申请人数,
    COUNT(CASE WHEN sex = 'Male' AND acception = 'Accepted' THEN 1 END) AS 男生录取人数,
    ROUND(COUNT(CASE WHEN sex = 'Male' AND acception = 'Accepted' THEN 1 END) / COUNT(CASE WHEN sex = 'Male' THEN 1 END) * 100, 2) AS 男生录取率,
    COUNT(*) AS 合计申请人数,
    COUNT(CASE WHEN acception = 'Accepted' THEN 1 END) AS 合计录取人数,
    ROUND(COUNT(CASE WHEN acception = 'Accepted' THEN 1 END) / COUNT(*) * 100, 2) AS 合计录取率
FROM application
WHERE department = 'Business School'

UNION

SELECT 
    '法学院' AS 学院,
    COUNT(CASE WHEN sex = 'Female' THEN 1 END) AS 女生申请人数,
    COUNT(CASE WHEN sex = 'Female' AND acception = 'Accepted' THEN 1 END) AS 女生录取人数,
    ROUND(COUNT(CASE WHEN sex = 'Female' AND acception = 'Accepted' THEN 1 END) / COUNT(CASE WHEN sex = 'Female' THEN 1 END) * 100, 2) AS 女生录取率,
    COUNT(CASE WHEN sex = 'Male' THEN 1 END) AS 男生申请人数,
    COUNT(CASE WHEN sex = 'Male' AND acception = 'Accepted' THEN 1 END) AS 男生录取人数,
    ROUND(COUNT(CASE WHEN sex = 'Male' AND acception = 'Accepted' THEN 1 END) / COUNT(CASE WHEN sex = 'Male' THEN 1 END) * 100, 2) AS 男生录取率,
    COUNT(*) AS 合计申请人数,
    COUNT(CASE WHEN acception = 'Accepted' THEN 1 END) AS 合计录取人数,
    ROUND(COUNT(CASE WHEN acception = 'Accepted' THEN 1 END) / COUNT(*) * 100, 2) AS 合计录取率
FROM application
WHERE department = 'Law School'

UNION

SELECT 
    '合计' AS 学院,
    COUNT(CASE WHEN sex = 'Female' THEN 1 END) AS 女生申请人数,
    COUNT(CASE WHEN sex = 'Female' AND acception = 'Accepted' THEN 1 END) AS 女生录取人数,
    ROUND(COUNT(CASE WHEN sex = 'Female' AND acception = 'Accepted' THEN 1 END) / COUNT(CASE WHEN sex = 'Female' THEN 1 END) * 100, 2) AS 女生录取率,
    COUNT(CASE WHEN sex = 'Male' THEN 1 END) AS 男生申请人数,
    COUNT(CASE WHEN sex = 'Male' AND acception = 'Accepted' THEN 1 END) AS 男生录取人数,
    ROUND(COUNT(CASE WHEN sex = 'Male' AND acception = 'Accepted' THEN 1 END) / COUNT(CASE WHEN sex = 'Male' THEN 1 END) * 100, 2) AS 男生录取率,
    COUNT(*) AS 合计申请人数,
    COUNT(CASE WHEN acception = 'Accepted' THEN 1 END) AS 合计录取人数,
    ROUND(COUNT(CASE WHEN acception = 'Accepted' THEN 1 END) / COUNT(*) * 100, 2) AS 合计录取率
FROM application;

 * mysql://stu2000017781:***@162.105.146.37:43306
3 rows affected.


[('商学院', 100, 49, Decimal('49.00'), 20, 15, Decimal('75.00'), 120, 64, Decimal('53.33')),
 ('法学院', 20, 1, Decimal('5.00'), 100, 10, Decimal('10.00'), 120, 11, Decimal('9.17')),
 ('合计', 120, 50, Decimal('41.67'), 120, 25, Decimal('20.83'), 240, 75, Decimal('31.25'))]

可以看到，无论是商学院还是法学院,男生的录取率均高于女生的录取率，但是在整体上，女生的录取率高于男生的录取率，这说明**存在**辛普森悖论。

#### 任务二：
下面计算条件概率，判断是否存在辛普森悖论

In [70]:
%%sql

# P(yes│男生)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Male') AS P_yes_male
FROM application
WHERE sex = 'Male' AND acception = 'Accepted';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.2083'),)]

In [71]:
%%sql

# P(yes│女生)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Female') AS P_yes_female
FROM application
WHERE sex = 'Female' AND acception = 'Accepted';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.4167'),)]

In [72]:
%%sql

# P(no│男生)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Male') AS P_no_male
FROM application
WHERE sex = 'Male' AND acception = 'Rejected';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.7917'),)]

In [73]:
%%sql

# P(no|女生)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Female') AS P_no_female
FROM application
WHERE sex = 'Female' AND acception = 'Rejected';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.5833'),)]

In [74]:
%%sql

# P(yes│<男生,商学院>)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Male' AND department = 'Business school') AS P_yes_male_business
FROM application
WHERE sex = 'Male' AND department = 'Business school' AND acception = 'Accepted';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.7500'),)]

In [75]:
%%sql

# P(yes│<女生,商学院>)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Female' AND department = 'Business school') AS P_yes_female_business
FROM application
WHERE sex = 'Female' AND department = 'Business school' AND acception = 'Accepted';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.4900'),)]

In [76]:
%%sql

# P(yes│<男生,法学院>)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Male' AND department = 'Law school') AS P_yes_female_business
FROM application
WHERE sex = 'Male' AND department = 'Law school' AND acception = 'Accepted';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.1000'),)]

In [77]:
%%sql

# P(yes│<女生,法学院>)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Female' AND department = 'Law school') AS P_yes_female_law
FROM application
WHERE sex = 'Female' AND department = 'Law school' AND acception = 'Accepted';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.0500'),)]

In [78]:
%%sql

# P(no│<男生,商学院>)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Male' AND department = 'Business school') AS P_yes_female_law
FROM application
WHERE sex = 'Male' AND department = 'Business school' AND acception = 'Rejected';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.2500'),)]

In [79]:
%%sql

# P(no│<女生,商学院>) 
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Female' AND department = 'Business school') AS P_no_female_business
FROM application
WHERE sex = 'Female' AND department = 'Business school' AND acception = 'Rejected';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.5100'),)]

In [80]:
%%sql

# P(no│<男生,法学院>)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Male' AND department = 'Law school') AS P_no_male_law
FROM application
WHERE sex = 'Male' AND department = 'Law school' AND acception = 'Rejected';


 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.9000'),)]

In [81]:
%%sql

# P(no│<女生,法学院>)
SELECT COUNT(*) / (SELECT COUNT(*) FROM application WHERE sex = 'Female' AND department = 'Law school') AS P_no_female_law
FROM application
WHERE sex = 'Female' AND department = 'Law school' AND acception = 'Rejected';

 * mysql://stu2000017781:***@162.105.146.37:43306
1 rows affected.


[(Decimal('0.9500'),)]

由计算结果我们可以再次看到，P(yes│<男生,商学院>)高于P(yes│<女生,商学院>)，P(yes│<男生,法学院>)高于P(yes│<女生,法学院>)，但是P(yes│<男生>)却低于P(yes│<女生>)，这说明**存在**辛普森悖论。当然我们把yes换成no，P(no│<男生,商学院>)低于P(no│<女生,商学院>)，P(no│<男生,法学院>)低于P(no│<女生,法学院>)，但是P(no│<男生>)却高于P(no│<女生>)，这也说明**存在**辛普森悖论。

### 练习二：KNN分类
这里我们使用威斯康辛乳腺癌数据集

#### 任务一：属性值的预处理


In [82]:
%%sql

# 首先需要在数据库中建表
DROP TABLE IF EXISTS breast_cancer;

CREATE TABLE breast_cancer (
   id INTEGER PRIMARY KEY,
   radius_mean REAL,
   texture_mean REAL,
   perimeter_mean REAL,
   area_mean REAL,
   smoothness_mean REAL,
   compactness_mean REAL,
   concavity_mean REAL,
   concave_points_mean REAL,
   symmetry_mean REAL,
   fractal_dimension_mean REAL,
   radius_se REAL,
   texture_se REAL,
   perimeter_se REAL,
   area_se REAL,
   smoothness_se REAL,
   compactness_se REAL,
   concavity_se REAL,
   concave_points_se REAL,
   symmetry_se REAL,
   fractal_dimension_se REAL,
   radius_worst REAL,
   texture_worst REAL,
   perimeter_worst REAL,
   area_worst REAL,
   smoothness_worst REAL,
   compactness_worst REAL,
   concavity_worst REAL,
   concave_points_worst REAL,
   symmetry_worst REAL,
   fractal_dimension_worst REAL,
   diagnosis BOOLEAN
);

 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.
0 rows affected.


[]

In [83]:
user_name = 'stu2000017781'
password = 'stu2000017781'
db_name = 'stu2000017781'

db = pymysql.connect(host='162.105.146.37',user=user_name, password=password,
port=43306,db=db_name)
cursor = db.cursor()

In [84]:
#导入乳腺癌数据
from sklearn.datasets import load_breast_cancer
import pandas as pd
import numpy as np

data = load_breast_cancer()

x = data.data
y = data.target
# name = ['平均半径','平均纹理','平均周长','平均面积','平均光滑度',
#         '平均紧凑度','平均凹度','平均凹点','平均对称','平均分形维数',
#         '半径误差','纹理误差','周长误差','面积误差','平滑度误差',
#         '紧凑度误差','凹度误差','凹点误差','对称误差','分形维数误差',
#         '最差半径','最差纹理','最差的边界','最差的区域','最差的平滑度',
#         '最差的紧凑性','最差的凹陷','最差的凹点','最差的对称性','最差的分形维数',
#         '患病否']
insert_query="INSERT INTO breast_cancer (id,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave_points_mean, symmetry_mean,fractal_dimension_mean,radius_se,texture_se,perimeter_se,area_se,smoothness_se,compactness_se,concavity_se,concave_points_se,symmetry_se,fractal_dimension_se,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave_points_worst,symmetry_worst,fractal_dimension_worst,diagnosis) \
 VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)"
for i in range(x.shape[0]):
    values=(i, x[i][0],x[i][1],x[i][2],x[i][3],x[i][4],x[i][5],x[i][6],x[i][7],x[i][8],x[i][9],x[i][10],x[i][11],x[i][12],x[i][13],x[i][14],x[i][15],x[i][16],x[i][17],x[i][18],x[i][19],x[i][20],x[i][21],x[i][22],x[i][23],x[i][24],x[i][25],x[i][26],x[i][27],x[i][28],x[i][29],y[i])
    try:
        cursor.execute(insert_query, values)
        db.commit()
    except:
        db.rollback()

In [85]:
%%sql

select * from breast_cancer limit 10;

 * mysql://stu2000017781:***@162.105.146.37:43306
10 rows affected.


[(0, 17.99, 10.38, 122.8, 1001.0, 0.1184, 0.2776, 0.3001, 0.1471, 0.2419, 0.07871, 1.095, 0.9053, 8.589, 153.4, 0.006399, 0.04904, 0.05373, 0.01587, 0.03003, 0.006193, 25.38, 17.33, 184.6, 2019.0, 0.1622, 0.6656, 0.7119, 0.2654, 0.4601, 0.1189, 0),
 (1, 20.57, 17.77, 132.9, 1326.0, 0.08474, 0.07864, 0.0869, 0.07017, 0.1812, 0.05667, 0.5435, 0.7339, 3.398, 74.08, 0.005225, 0.01308, 0.0186, 0.0134, 0.01389, 0.003532, 24.99, 23.41, 158.8, 1956.0, 0.1238, 0.1866, 0.2416, 0.186, 0.275, 0.08902, 0),
 (2, 19.69, 21.25, 130.0, 1203.0, 0.1096, 0.1599, 0.1974, 0.1279, 0.2069, 0.05999, 0.7456, 0.7869, 4.585, 94.03, 0.00615, 0.04006, 0.03832, 0.02058, 0.0225, 0.004571, 23.57, 25.53, 152.5, 1709.0, 0.1444, 0.4245, 0.4504, 0.243, 0.3613, 0.08758, 0),
 (3, 11.42, 20.38, 77.58, 386.1, 0.1425, 0.2839, 0.2414, 0.1052, 0.2597, 0.09744, 0.4956, 1.156, 3.445, 27.23, 0.00911, 0.07458, 0.05661, 0.01867, 0.05963, 0.009208, 14.91, 26.5, 98.87, 567.7, 0.2098, 0.8663, 0.6869, 0.2575, 0.6638, 0.173, 0),
 (4, 20.29, 14.34, 135.1, 1297.0, 0.1003, 0.1328, 0.198, 0.1043, 0.1809, 0.05883, 0.7572, 0.7813, 5.438, 94.44, 0.01149, 0.02461, 0.05688, 0.01885, 0.01756, 0.005115, 22.54, 16.67, 152.2, 1575.0, 0.1374, 0.205, 0.4, 0.1625, 0.2364, 0.07678, 0),
 (5, 12.45, 15.7, 82.57, 477.1, 0.1278, 0.17, 0.1578, 0.08089, 0.2087, 0.07613, 0.3345, 0.8902, 2.217, 27.19, 0.00751, 0.03345, 0.03672, 0.01137, 0.02165, 0.005082, 15.47, 23.75, 103.4, 741.6, 0.1791, 0.5249, 0.5355, 0.1741, 0.3985, 0.1244, 0),
 (6, 18.25, 19.98, 119.6, 1040.0, 0.09463, 0.109, 0.1127, 0.074, 0.1794, 0.05742, 0.4467, 0.7732, 3.18, 53.91, 0.004314, 0.01382, 0.02254, 0.01039, 0.01369, 0.002179, 22.88, 27.66, 153.2, 1606.0, 0.1442, 0.2576, 0.3784, 0.1932, 0.3063, 0.08368, 0),
 (7, 13.71, 20.83, 90.2, 577.9, 0.1189, 0.1645, 0.09366, 0.05985, 0.2196, 0.07451, 0.5835, 1.377, 3.856, 50.96, 0.008805, 0.03029, 0.02488, 0.01448, 0.01486, 0.005412, 17.06, 28.14, 110.6, 897.0, 0.1654, 0.3682, 0.2678, 0.1556, 0.3196, 0.1151, 0),
 (8, 13.0, 21.82, 87.5, 519.8, 0.1273, 0.1932, 0.1859, 0.09353, 0.235, 0.07389, 0.3063, 1.002, 2.406, 24.32, 0.005731, 0.03502, 0.03553, 0.01226, 0.02143, 0.003749, 15.49, 30.73, 106.2, 739.3, 0.1703, 0.5401, 0.539, 0.206, 0.4378, 0.1072, 0),
 (9, 12.46, 24.04, 83.97, 475.9, 0.1186, 0.2396, 0.2273, 0.08543, 0.203, 0.08243, 0.2976, 1.599, 2.039, 23.94, 0.007149, 0.07217, 0.07743, 0.01432, 0.01789, 0.01008, 15.09, 40.68, 97.65, 711.4, 0.1853, 1.058, 1.105, 0.221, 0.4366, 0.2075, 0)]

为了将除了id与diagnosis之外的属性值转化为[0, 1]之间的数值，另一方面数据的极差相较于数据于最小值的差值不大，因此我们使用最小-最大规范化。

In [86]:
%%sql

SET @min_radius_mean = (SELECT MIN(radius_mean) from breast_cancer);
SET @max_radius_mean = (SELECT MAX(radius_mean) from breast_cancer);
SET @min_texture_mean = (SELECT MIN(texture_mean) from breast_cancer);
SET @max_texture_mean = (SELECT MAX(texture_mean) from breast_cancer);
SET @min_perimeter_mean = (SELECT MIN(perimeter_mean) from breast_cancer);
SET @max_perimeter_mean = (SELECT MAX(perimeter_mean) from breast_cancer);
SET @min_area_mean = (SELECT MIN(area_mean) from breast_cancer);
SET @max_area_mean = (SELECT MAX(area_mean) from breast_cancer);
SET @min_smoothness_mean = (SELECT MIN(smoothness_mean) from breast_cancer);
SET @max_smoothness_mean = (SELECT MAX(smoothness_mean) from breast_cancer);
SET @min_compactness_mean = (SELECT MIN(compactness_mean) from breast_cancer);
SET @max_compactness_mean = (SELECT MAX(compactness_mean) from breast_cancer);
SET @min_concavity_mean = (SELECT MIN(concavity_mean) from breast_cancer);
SET @max_concavity_mean = (SELECT MAX(concavity_mean) from breast_cancer);
SET @min_concave_points_mean = (SELECT MIN(concave_points_mean) from breast_cancer);
SET @max_concave_points_mean = (SELECT MAX(concave_points_mean) from breast_cancer);
SET @min_symmetry_mean = (SELECT MIN(symmetry_mean) from breast_cancer);
SET @max_symmetry_mean = (SELECT MAX(symmetry_mean) from breast_cancer);
SET @min_fractal_dimension_mean = (SELECT MIN(fractal_dimension_mean) from breast_cancer);
SET @max_fractal_dimension_mean = (SELECT MAX(fractal_dimension_mean) from breast_cancer);
SET @min_radius_se = (SELECT MIN(radius_se) from breast_cancer);
SET @max_radius_se = (SELECT MAX(radius_se) from breast_cancer);
SET @min_texture_se = (SELECT MIN(texture_se) from breast_cancer);
SET @max_texture_se = (SELECT MAX(texture_se) from breast_cancer);
SET @min_perimeter_se = (SELECT MIN(perimeter_se) from breast_cancer);
SET @max_perimeter_se = (SELECT MAX(perimeter_se) from breast_cancer);
SET @min_area_se = (SELECT MIN(area_se) from breast_cancer);
SET @max_area_se = (SELECT MAX(area_se) from breast_cancer);
SET @min_smoothness_se = (SELECT MIN(smoothness_se) from breast_cancer);
SET @max_smoothness_se = (SELECT MAX(smoothness_se) from breast_cancer);
SET @min_compactness_se = (SELECT MIN(compactness_se) from breast_cancer);
SET @max_compactness_se = (SELECT MAX(compactness_se) from breast_cancer);
SET @min_concavity_se = (SELECT MIN(concavity_se) from breast_cancer);
SET @max_concavity_se = (SELECT MAX(concavity_se) from breast_cancer);
SET @min_concave_points_se = (SELECT MIN(concave_points_se) from breast_cancer);
SET @max_concave_points_se = (SELECT MAX(concave_points_se) from breast_cancer);
SET @min_symmetry_se = (SELECT MIN(symmetry_se) from breast_cancer);
SET @max_symmetry_se = (SELECT MAX(symmetry_se) from breast_cancer);
SET @min_fractal_dimension_se = (SELECT MIN(fractal_dimension_se) from breast_cancer);
SET @max_fractal_dimension_se = (SELECT MAX(fractal_dimension_se) from breast_cancer);
SET @min_radius_worst = (SELECT MIN(radius_worst) from breast_cancer);
SET @max_radius_worst = (SELECT MAX(radius_worst) from breast_cancer);
SET @min_texture_worst = (SELECT MIN(texture_worst) from breast_cancer);
SET @max_texture_worst = (SELECT MAX(texture_worst) from breast_cancer);
SET @min_perimeter_worst = (SELECT MIN(perimeter_worst) from breast_cancer);
SET @max_perimeter_worst = (SELECT MAX(perimeter_worst) from breast_cancer);
SET @min_area_worst = (SELECT MIN(area_worst) from breast_cancer);
SET @max_area_worst = (SELECT MAX(area_worst) from breast_cancer);
SET @min_smoothness_worst = (SELECT MIN(smoothness_worst) from breast_cancer);
SET @max_smoothness_worst = (SELECT MAX(smoothness_worst) from breast_cancer);
SET @min_compactness_worst = (SELECT MIN(compactness_worst) from breast_cancer);
SET @max_compactness_worst = (SELECT MAX(compactness_worst) from breast_cancer);
SET @min_concavity_worst = (SELECT MIN(concavity_worst) from breast_cancer);
SET @max_concavity_worst = (SELECT MAX(concavity_worst) from breast_cancer);
SET @min_concave_points_worst = (SELECT MIN(concave_points_worst) from breast_cancer);
SET @max_concave_points_worst = (SELECT MAX(concave_points_worst) from breast_cancer);
SET @min_symmetry_worst = (SELECT MIN(symmetry_worst) from breast_cancer);
SET @max_symmetry_worst = (SELECT MAX(symmetry_worst) from breast_cancer);
SET @min_fractal_dimension_worst = (SELECT MIN(fractal_dimension_worst) from breast_cancer);
SET @max_fractal_dimension_worst = (SELECT MAX(fractal_dimension_worst) from breast_cancer);


 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected

[]

In [87]:
%%sql

UPDATE breast_cancer
SET 
    radius_mean = (radius_mean - @min_radius_mean) / (@max_radius_mean - @min_radius_mean),
    texture_mean = (texture_mean - @min_texture_mean) / (@max_texture_mean - @min_texture_mean),
    perimeter_mean = (perimeter_mean - @min_perimeter_mean) / (@max_perimeter_mean - @min_perimeter_mean),
    area_mean = (area_mean - @min_area_mean) / (@max_area_mean - @min_area_mean),
    smoothness_mean = (smoothness_mean - @min_smoothness_mean) / (@max_smoothness_mean - @min_smoothness_mean),
    compactness_mean = (compactness_mean - @min_compactness_mean) / (@max_compactness_mean - @min_compactness_mean),
    concavity_mean = (concavity_mean - @min_concavity_mean) / (@max_concavity_mean - @min_concavity_mean),
    concave_points_mean = (concave_points_mean - @min_concave_points_mean) / (@max_concave_points_mean - @min_concave_points_mean),
    symmetry_mean = (symmetry_mean - @min_symmetry_mean) / (@max_symmetry_mean - @min_symmetry_mean),
    fractal_dimension_mean = (fractal_dimension_mean - @min_fractal_dimension_mean) / (@max_fractal_dimension_mean - @min_fractal_dimension_mean),
    radius_se = (radius_se - @min_radius_se) / (@max_radius_se - @min_radius_se),
    texture_se = (texture_se - @min_texture_se) / (@max_texture_se - @min_texture_se),
    perimeter_se = (perimeter_se - @min_perimeter_se) / (@max_perimeter_se - @min_perimeter_se),
    area_se = (area_se - @min_area_se) / (@max_area_se - @min_area_se),
    smoothness_se = (smoothness_se - @min_smoothness_se) / (@max_smoothness_se - @min_smoothness_se),
    compactness_se = (compactness_se - @min_compactness_se) / (@max_compactness_se - @min_compactness_se),
    concavity_se = (concavity_se - @min_concavity_se) / (@max_concavity_se - @min_concavity_se),
    concave_points_se = (concave_points_se - @min_concave_points_se) / (@max_concave_points_se - @min_concave_points_se),
    symmetry_se = (symmetry_se - @min_symmetry_se) / (@max_symmetry_se - @min_symmetry_se),
    fractal_dimension_se = (fractal_dimension_se - @min_fractal_dimension_se) / (@max_fractal_dimension_se - @min_fractal_dimension_se),
    radius_worst = (radius_worst - @min_radius_worst) / (@max_radius_worst - @min_radius_worst),
    texture_worst = (texture_worst - @min_texture_worst) / (@max_texture_worst - @min_texture_worst),
    perimeter_worst = (perimeter_worst - @min_perimeter_worst) / (@max_perimeter_worst - @min_perimeter_worst),
    area_worst = (area_worst - @min_area_worst) / (@max_area_worst - @min_area_worst),
    smoothness_worst = (smoothness_worst - @min_smoothness_worst) / (@max_smoothness_worst - @min_smoothness_worst),
    compactness_worst = (compactness_worst - @min_compactness_worst) / (@max_compactness_worst - @min_compactness_worst),
    concavity_worst = (concavity_worst - @min_concavity_worst) / (@max_concavity_worst - @min_concavity_worst),
    concave_points_worst = (concave_points_worst - @min_concave_points_worst) / (@max_concave_points_worst - @min_concave_points_worst),
    symmetry_worst = (symmetry_worst - @min_symmetry_worst) / (@max_symmetry_worst - @min_symmetry_worst),
    fractal_dimension_worst = (fractal_dimension_worst - @min_fractal_dimension_worst) / (@max_fractal_dimension_worst - @min_fractal_dimension_worst);

 * mysql://stu2000017781:***@162.105.146.37:43306
569 rows affected.


[]

In [88]:
%%sql

select * from breast_cancer limit 10;

 * mysql://stu2000017781:***@162.105.146.37:43306
10 rows affected.


[(0, 0.5210374366983767, 0.022658099425092997, 0.5459885287817012, 0.3637327677624602, 0.5937528211609642, 0.7920372983252562, 0.7031396438612932, 0.7311133200795229, 0.6863636363636364, 0.6055181128896379, 0.35614702154626104, 0.12046941301272983, 0.3690335956273854, 0.2738112581668217, 0.15929564537512322, 0.3513984438369334, 0.13568181818181818, 0.30062511839363515, 0.3116451848933416, 0.1830424388154167, 0.6207755247242974, 0.1415245202558635, 0.6683101748095025, 0.4506979944946913, 0.6011358383411478, 0.6192915563058473, 0.5686102236421725, 0.9120274914089348, 0.5984624482554701, 0.4188639643185098, 0),
 (1, 0.6431444933503716, 0.2725735542779844, 0.6157832907193699, 0.5015906680805938, 0.28987993138936535, 0.18176798969388383, 0.2036082474226804, 0.34875745526838964, 0.3797979797979798, 0.14132266217354678, 0.15643671917436175, 0.08258928571428571, 0.12444046553267682, 0.12565978953974424, 0.11938674915864973, 0.0813230390242437, 0.04696969696969697, 0.2538359537791249, 0.0845387516181685, 0.09111009770186422, 0.6069014585556741, 0.3035714285714286, 0.5398177200059765, 0.43521431380259534, 0.34755332496863234, 0.15456335923780692, 0.1929712460063898, 0.6391752577319588, 0.2335895919574217, 0.2228781319690411, 0),
 (2, 0.6014955748024045, 0.39026039905309434, 0.5957432105590491, 0.4494167550371156, 0.5143089284102195, 0.4310165020550886, 0.46251171508903466, 0.6356858846918491, 0.5095959595959596, 0.21124684077506325, 0.22962158247329348, 0.09430251060820369, 0.18037035291900297, 0.16292178902423993, 0.15083115205493422, 0.28395469702887016, 0.09676767676767677, 0.38984656184883504, 0.20569032475938534, 0.1270055138675843, 0.5563856278904304, 0.3600746268656717, 0.5084416554609293, 0.3745084545812033, 0.48358977745492965, 0.38537512976491933, 0.35974440894568693, 0.8350515463917526, 0.40370589394835404, 0.21343303161484986, 0),
 (3, 0.21009039708457572, 0.36083868785931683, 0.23350148573008084, 0.10290562036055144, 0.8113207547169811, 0.8113612661799889, 0.5656044985941893, 0.5228628230616302, 0.7762626262626262, 1.0, 0.13909107369183413, 0.1758751768033946, 0.126655044055977, 0.03815479325660537, 0.2514532413230445, 0.5432150689458347, 0.14295454545454545, 0.3536654669444971, 0.7281476895367817, 0.28720478697676993, 0.24831020988971897, 0.3859275053304905, 0.2413466806115843, 0.09400806134486828, 0.9154724955424947, 0.814011700672352, 0.5486421725239616, 0.8848797250859107, 1.0, 0.773711137347501, 0),
 (4, 0.6298925647214729, 0.15657761244504562, 0.6309861101513371, 0.4892895015906681, 0.43035117811681867, 0.34789276731488866, 0.4639175257731959, 0.5183896620278331, 0.37828282828282833, 0.18681550126369004, 0.23382219808075316, 0.09306488684582744, 0.22056259718230223, 0.16368757447730473, 0.3323588401264575, 0.16791840658515336, 0.14363636363636362, 0.3570752036370525, 0.13617943378173014, 0.1457996489918881, 0.5197438633938101, 0.12393390191897662, 0.5069475571492604, 0.34157491152182456, 0.4373637984547315, 0.1724151313172473, 0.31948881789137384, 0.5584192439862543, 0.15750049280504633, 0.14259477895841535, 0),
 (5, 0.25883856311231007, 0.20257017247210005, 0.26798424435077045, 0.14150583244962886, 0.6786133429628961, 0.4619961965523588, 0.3697282099343955, 0.402037773359841, 0.5186868686868686, 0.5511794439764113, 0.08075321383306175, 0.11713224893917965, 0.06879329029826133, 0.0380800824806966, 0.19706292279974164, 0.23431069186168776, 0.09272727272727273, 0.21538170107974997, 0.19372994878144872, 0.14465956358912704, 0.2682319459267165, 0.312633262260128, 0.2639075651177848, 0.13674793550924108, 0.7127385590701975, 0.48278371219838756, 0.42771565495207664, 0.5982817869415809, 0.47703528484131685, 0.4549390003935459, 0),
 (6, 0.5333427989966397, 0.3473114643219479, 0.5238753368806579, 0.3802757158006363, 0.3791640335831002, 0.27489111097478686, 0.26405810684161196, 0.36779324055666, 0.3707070707070707, 0.1571187868576243, 0.12138330617418068, 0.09127475247524752, 0.11416859067992273, 0.08798688078775041, 0.08841826

#### 任务二：数据集的划分

In [89]:
%%sql

ALTER TABLE breast_cancer
ADD COLUMN random_number DOUBLE;

UPDATE breast_cancer
SET random_number = RAND();

 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.
569 rows affected.


[]

In [90]:
%%sql

ALTER TABLE breast_cancer
ADD COLUMN is_train BOOLEAN;

UPDATE breast_cancer
SET is_train = IF(random_number < 0.7, TRUE, FALSE);

 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.
569 rows affected.


[]

#### 任务三：实现KNN算法

In [91]:
%%sql

DROP TABLE IF EXISTS breast_cancer_distance;

CREATE TABLE breast_cancer_distance(
    id_from INT,
    id_to INT,
    distance DOUBLE,
    is_diagnosis BOOLEAN
)

 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.
0 rows affected.


[]

In [92]:
%%sql

# 计算测试集与训练集之间的距离
INSERT INTO breast_cancer_distance(id_from, id_to, distance, is_diagnosis)
SELECT bc1.id, bc2.id, SQRT(POW(bc1.radius_mean-bc2.radius_mean, 2) + POW(bc1.texture_mean-bc2.texture_mean, 2) + POW(bc1.perimeter_mean-bc2.perimeter_mean, 2) + 
    POW(bc1.area_mean-bc2.area_mean, 2) + POW(bc1.smoothness_mean-bc2.smoothness_mean, 2) + POW(bc1.compactness_mean-bc2.compactness_mean, 2) + POW(bc1.concavity_mean-bc2.concavity_mean, 2) 
    + POW(bc1.concave_points_mean-bc2.concave_points_mean, 2) + POW(bc1.symmetry_mean-bc2.symmetry_mean, 2) + POW(bc1.fractal_dimension_mean-bc2.fractal_dimension_mean, 2) + POW(bc1.radius_se-bc2.radius_se, 2)
    + POW(bc1.texture_se-bc2.texture_se, 2) + POW(bc1.perimeter_se-bc2.perimeter_se, 2) + POW(bc1.area_se-bc2.area_se, 2) + POW(bc1.smoothness_se-bc2.smoothness_se, 2) + POW(bc1.compactness_se-bc2.compactness_se, 2) + 
    POW(bc1.concavity_se-bc2.concavity_se, 2) + POW(bc1.concave_points_se-bc2.concave_points_se, 2) + POW(bc1.symmetry_se-bc2.symmetry_se, 2) + POW(bc1.fractal_dimension_se-bc2.fractal_dimension_se, 2) 
    + POW(bc1.radius_worst-bc2.radius_worst, 2) + POW(bc1.texture_worst-bc2.texture_worst, 2) + POW(bc1.perimeter_worst-bc2.perimeter_worst, 2) + POW(bc1.area_worst-bc2.area_worst, 2) + POW(bc1.smoothness_worst-bc2.smoothness_worst, 2) 
    + POW(bc1.compactness_worst-bc2.compactness_worst, 2) + POW(bc1.concavity_worst-bc2.concavity_worst, 2) + POW(bc1.concave_points_worst-bc2.concave_points_worst, 2) + POW(bc1.symmetry_worst-bc2.symmetry_worst, 2) + POW(bc1.fractal_dimension_worst-bc2.fractal_dimension_worst, 2) ) AS distance, bc2.diagnosis
FROM breast_cancer AS bc1, breast_cancer AS bc2
WHERE (NOT bc1.is_train) AND bc2.is_train

 * mysql://stu2000017781:***@162.105.146.37:43306
66420 rows affected.


[]

In [93]:
%%sql

SELECT * FROM breast_cancer_distance LIMIT 10;

 * mysql://stu2000017781:***@162.105.146.37:43306
10 rows affected.


[(551, 3, 2.0615665374362266, 0),
 (546, 3, 2.3673580363398194, 0),
 (541, 3, 1.797292671285451, 0),
 (537, 3, 1.7014808802258243, 0),
 (532, 3, 2.2821060246488094, 0),
 (531, 3, 1.9791727527006795, 0),
 (529, 3, 2.107046237852332, 0),
 (527, 3, 2.2488835842504864, 0),
 (526, 3, 1.9048131953122636, 0),
 (523, 3, 1.9123822621797186, 0)]

这里比较难受的一个点是因为当前的MySQL Server不支持LIMIT字句后面跟随一个变量， 因此我们后续选取K值只能手动选取。

In [94]:
%%sql

ALTER TABLE breast_cancer
ADD COLUMN predict_true BOOLEAN;

UPDATE breast_cancer
SET predict_true = IF(diagnosis = (SELECT is_diagnosis
FROM 
        (SELECT bcd.id_to, bcd.is_diagnosis
        FROM breast_cancer_distance AS bcd
        WHERE (breast_cancer.id = bcd.id_from)
        ORDER BY distance ASC
        LIMIT 13) as t
GROUP BY is_diagnosis
ORDER BY COUNT(*) DESC
LIMIT 1), TRUE, FALSE);

 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.
569 rows affected.


[]

In [95]:
%%sql

SELECT * FROM breast_cancer LIMIT 30;

 * mysql://stu2000017781:***@162.105.146.37:43306
30 rows affected.


[(0, 0.5210374366983767, 0.022658099425092997, 0.5459885287817012, 0.3637327677624602, 0.5937528211609642, 0.7920372983252562, 0.7031396438612932, 0.7311133200795229, 0.6863636363636364, 0.6055181128896379, 0.35614702154626104, 0.12046941301272983, 0.3690335956273854, 0.2738112581668217, 0.15929564537512322, 0.3513984438369334, 0.13568181818181818, 0.30062511839363515, 0.3116451848933416, 0.1830424388154167, 0.6207755247242974, 0.1415245202558635, 0.6683101748095025, 0.4506979944946913, 0.6011358383411478, 0.6192915563058473, 0.5686102236421725, 0.9120274914089348, 0.5984624482554701, 0.4188639643185098, 0, 0.9141323174481581, 0, 1),
 (1, 0.6431444933503716, 0.2725735542779844, 0.6157832907193699, 0.5015906680805938, 0.28987993138936535, 0.18176798969388383, 0.2036082474226804, 0.34875745526838964, 0.3797979797979798, 0.14132266217354678, 0.15643671917436175, 0.08258928571428571, 0.12444046553267682, 0.12565978953974424, 0.11938674915864973, 0.0813230390242437, 0.04696969696969697, 0.2538359537791249, 0.0845387516181685, 0.09111009770186422, 0.6069014585556741, 0.3035714285714286, 0.5398177200059765, 0.43521431380259534, 0.34755332496863234, 0.15456335923780692, 0.1929712460063898, 0.6391752577319588, 0.2335895919574217, 0.2228781319690411, 0, 0.71518206290452, 0, 1),
 (2, 0.6014955748024045, 0.39026039905309434, 0.5957432105590491, 0.4494167550371156, 0.5143089284102195, 0.4310165020550886, 0.46251171508903466, 0.6356858846918491, 0.5095959595959596, 0.21124684077506325, 0.22962158247329348, 0.09430251060820369, 0.18037035291900297, 0.16292178902423993, 0.15083115205493422, 0.28395469702887016, 0.09676767676767677, 0.38984656184883504, 0.20569032475938534, 0.1270055138675843, 0.5563856278904304, 0.3600746268656717, 0.5084416554609293, 0.3745084545812033, 0.48358977745492965, 0.38537512976491933, 0.35974440894568693, 0.8350515463917526, 0.40370589394835404, 0.21343303161484986, 0, 0.8335133929117707, 0, 1),
 (3, 0.21009039708457572, 0.36083868785931683, 0.23350148573008084, 0.10290562036055144, 0.8113207547169811, 0.8113612661799889, 0.5656044985941893, 0.5228628230616302, 0.7762626262626262, 1.0, 0.13909107369183413, 0.1758751768033946, 0.126655044055977, 0.03815479325660537, 0.2514532413230445, 0.5432150689458347, 0.14295454545454545, 0.3536654669444971, 0.7281476895367817, 0.28720478697676993, 0.24831020988971897, 0.3859275053304905, 0.2413466806115843, 0.09400806134486828, 0.9154724955424947, 0.814011700672352, 0.5486421725239616, 0.8848797250859107, 1.0, 0.773711137347501, 0, 0.022020806578938668, 1, 0),
 (4, 0.6298925647214729, 0.15657761244504562, 0.6309861101513371, 0.4892895015906681, 0.43035117811681867, 0.34789276731488866, 0.4639175257731959, 0.5183896620278331, 0.37828282828282833, 0.18681550126369004, 0.23382219808075316, 0.09306488684582744, 0.22056259718230223, 0.16368757447730473, 0.3323588401264575, 0.16791840658515336, 0.14363636363636362, 0.3570752036370525, 0.13617943378173014, 0.1457996489918881, 0.5197438633938101, 0.12393390191897662, 0.5069475571492604, 0.34157491152182456, 0.4373637984547315, 0.1724151313172473, 0.31948881789137384, 0.5584192439862543, 0.15750049280504633, 0.14259477895841535, 0, 0.609563884893026, 1, 0),
 (5, 0.25883856311231007, 0.20257017247210005, 0.26798424435077045, 0.14150583244962886, 0.6786133429628961, 0.4619961965523588, 0.3697282099343955, 0.402037773359841, 0.5186868686868686, 0.5511794439764113, 0.08075321383306175, 0.11713224893917965, 0.06879329029826133, 0.0380800824806966, 0.19706292279974164, 0.23431069186168776, 0.09272727272727273, 0.21538170107974997, 0.19372994878144872, 0.14465956358912704, 0.2682319459267165, 0.312633262260128, 0.2639075651177848, 0.13674793550924108, 0.7127385590701975, 0.48278371219838756, 0.42771565495207664, 0.5982817869415809, 0.47703528484131685, 0.4549390003935459, 0, 0.9817570354619595, 0, 1),
 (6, 0.5333427989966397, 0.3473114643219479, 0.5238753368806579, 0.3802757158006363, 0.3791640335831002, 0.27489111097478686, 0.2640581068416119

In [96]:
%%sql

SET @num_sum = (SELECT COUNT(*) FROM breast_cancer WHERE NOT is_train);
SET @num_true = (SELECT COUNT(*) FROM breast_cancer WHERE predict_true AND (NOT is_train));

SELECT @num_true / @num_sum;

 * mysql://stu2000017781:***@162.105.146.37:43306
0 rows affected.
0 rows affected.
1 rows affected.


[(Decimal('0.9634'),)]

#### 任务四：讨论最佳K
下面我们将实验不同k值的结果，找出最优的k值。
正如前文所提到的，由于MySQL Server不支持LIMIT字句后面跟随一个变量，因此我们只能手动选取k值。为了节约篇幅，我们直接给出实验结果而不修改代码（实际上代码修改很简单，只需要修改LIMIT之后的常数即可）。
在knn中，我们一般选择k值为奇数来避免出现平票的情况，另一方面经验认为k=5效果比较好，综上我们选取了1,3,5,7,9,11,13这七个k值进行实验。

| K | 1 | 3 | 5 | 7 | 9 | 11 | 13 |
| :----: | :----: | :----: | :----: | :----: | :----: | :----: | :----: |
|Acurracy| 0.9630 | 0.9809 | 0.9706 | 0.9778 | 0.9632 | 0.9689 | 0.9634 |

可以看到在威斯康辛乳腺癌数据集上，当k=3时，准确率最高，为0.9809。

#### 

#### 与scikit-learn的结果对比
下面我们用scikit-learn中的KNN算法进行实验，与我们自己实现的KNN算法进行对比。

In [97]:
name = ['平均半径','平均纹理','平均周长','平均面积','平均光滑度',
        '平均紧凑度','平均凹度','平均凹点','平均对称','平均分形维数',
        '半径误差','纹理误差','周长误差','面积误差','平滑度误差',
        '紧凑度误差','凹度误差','凹点误差','对称误差','分形维数误差',
        '最差半径','最差纹理','最差的边界','最差的区域','最差的平滑度',
        '最差的紧凑性','最差的凹陷','最差的凹点','最差的对称性','最差的分形维数',
        '患病否']

breast_cancer = pd.DataFrame(np.concatenate((x,y.reshape(-1,1)),axis=1),columns=name)
breast_cancer.head()

平均半径   平均纹理    平均周长    平均面积    平均光滑度    平均紧凑度    平均凹度     平均凹点    平均对称  \
0  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.3001  0.14710  0.2419   
1  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.0869  0.07017  0.1812   
2  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.1974  0.12790  0.2069   
3  11.42  20.38   77.58   386.1  0.14250  0.28390  0.2414  0.10520  0.2597   
4  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.1980  0.10430  0.1809   

    平均分形维数  ...   最差纹理   最差的边界   最差的区域  最差的平滑度  最差的紧凑性   最差的凹陷   最差的凹点  \
0  0.07871  ...  17.33  184.60  2019.0  0.1622  0.6656  0.7119  0.2654   
1  0.05667  ...  23.41  158.80  1956.0  0.1238  0.1866  0.2416  0.1860   
2  0.05999  ...  25.53  152.50  1709.0  0.1444  0.4245  0.4504  0.2430   
3  0.09744  ...  26.50   98.87   567.7  0.2098  0.8663  0.6869  0.2575   
4  0.05883  ...  16.67  152.20  1575.0  0.1374  0.2050  0.4000  0.1625   

   最差的对称性  最差的分形维数  患病否  
0  0.4601  0.11890  0.0  
1  0.2750  0.08902  0.0  
2  0.3613  0.08758  0.0  
3  0.6638  0.17300  0.0  
4  0.2364  0.07678  0.0  

[5 rows x 31 columns]

In [98]:
#划分训练集和测试集
from sklearn.model_selection import train_test_split 
xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.3,random_state=420)

In [99]:
#归一化
from sklearn.preprocessing import MinMaxScaler as mms

mms_01 = mms().fit(xtrain) #求训练集最大/最小值
mms_02 = mms().fit(xtest) #求测试集最大/最小值

#转化
x_train = mms_01.transform(xtrain)
x_test = mms_02.transform(xtest)

In [104]:
#导入包
from sklearn.neighbors import KNeighborsClassifier

#建立模型
neighbors = [1, 3, 5, 7, 9, 11, 13]
for nn in neighbors:
    clf = KNeighborsClassifier(n_neighbors=nn)
    clf = clf.fit(x_train,ytrain)
    print(clf.score(x_test,ytest))


0.9415204678362573
0.9649122807017544
0.9766081871345029
0.9766081871345029
0.9766081871345029
0.9649122807017544
0.9707602339181286


scikit-learn中的KNN算法的结果如下：

| K | 1 | 3 | 5 | 7 | 9 | 11 | 13 |
| :----: | :----: | :----: | :----: | :----: | :----: | :----: | :----: |
|Acurracy| 0.9415 | 0.9649 | 0.9766 | 0.9766 | 0.9766 | 0.9649 | 0.9707 |

可以看到在威斯康辛乳腺癌数据集上，当k=5或7时，准确率最高，为0.9766。
总体来说，无论是scikit-learn中的knn算法还是我们自己使用SQL实现的knn,在预测的准确率上都十分相近，均位于0.96-0.98之间。这在一定程度上说明了我们自己用SQL实现的knn算法的正确性。